### PR TITLE
Fix AttributeError: 'SendCampaignThread' object has no attribute 'campaign'

### PR DESCRIPTION
**Pull request by danhayden**
addresses https://github.com/neon-jungle/wagtail-birdsong/issues/26

### CODE REVIEW

File: tests/app/settings.py [hunk 1]
```diff
@@ -91,3 +91,5 @@
 
 MEDIA_ROOT = os.path.join(os.path.dirname(__file__), 'media')
 MEDIA_URL = '/media/'
+
+WAGTAILADMIN_BASE_URL = "localhost:8080"
```
Comment (danhayden) [inline tests/app/settings.py:95]: Added this to suppress the following warning when running test:

```bash
(wagtailadmin.W003) The WAGTAILADMIN_BASE_URL setting is not defined
        HINT: This should be the base URL used to access the Wagtail admin site. Without this, URLs in notification emails will not display correctly.
```

File: tests/app/models.py [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
+from wagtail.admin.edit_handlers import FieldPanel
 from wagtail.core.fields import StreamField
 
 from birdsong.blocks import DefaultBlocks
 from birdsong.models import Campaign, Contact
 
 
 class SaleCampaign(Campaign):
-    body = StreamField(DefaultBlocks())
+    body = StreamField(DefaultBlocks(), use_json_field=True)
```
Comment (danhayden) [inline tests/app/models.py:10]: These changes suppress the following warning when running tests:

```bash
RemovedInWagtail50Warning: StreamFieldPanel is no longer required when using StreamField, and should be replaced by FieldPanel. StreamFieldPanel will be removed in a future release. See https://docs.wagtail.org/en/stable/releases/3.0.html#removal-of-special-purpose-field-panel-types
  StreamFieldPanel('body'),
```

**Comment by danhayden** [comment]
I'd like to add a test for handling an SMTPException however I'm new to writing Python/Django tests and do not know how to trigger an SMTPException such as a 429 HTTP response within a test.

Would you be able to offer some guidance?

Use the existing tests as a reference and base so far I have the following:

```python
class TestErrorHandling(WagtailTestUtils, TransactionTestCase):
    def setUp(self):
        self.campaign = SaleCampaign.objects.create(
            name='Test campaign',
            subject='The subject',
            body=[('rich_text', RichText('<p>The body</p>'))]
        )
        for person in [
            ('Terry', 'Testington', 'North', 'terry@tests.com'),
            ('Wag', 'Tail', 'South', 'wag@tail.com'),
        ]:
            ExtendedContact.objects.create(
                first_name=person[0],
                last_name=person[1],
                location=person[2],
                email=person[3],
            )
        self.login()


    def test_smtp_exception(self):
        self.client.get(f'/admin/app/salecampaign/send_campaign/{self.campaign.id}/')

        sleep(10)  # Allow time  to send
        self.assertEquals(len(mail.outbox), 0)
        self.assertEquals(self.campaign.receipts.all().count(), 0)
        # Get fresh from db (altered in a thread)
        fresh_campaign = SaleCampaign.objects.get(pk=self.campaign.pk)
        self.assertEquals(fresh_campaign.status, CampaignStatus.FAILED)
```

But this fails as no SMTPException occurs

```bash
Found 7 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
....F..
======================================================================
FAIL: test_smtp_exception (tests.test_admin.TestErrorHandling)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/tests/test_admin.py", line 134, in test_smtp_exception
    self.assertEquals(len(mail.outbox), 0)
AssertionError: 2 != 0

----------------------------------------------------------------------
Ran 7 tests in 38.539s

FAILED (failures=1)
Destroying test database for alias 'default'...
```

**Comment by david-smejkal** [comment]
Ideally you would simulate the situation of where the code throws such an exception in your test because you want to figure out whether the exception is an acceptable situation or whether it's something that should be remedied in some way.

Now to answer your question you can "throw" and exception forcefully with the `raise` keyword and then test for it as follows:
```python
from smtplib import SMTPException
...
def something_that_i_am_testing_that_raises_smtp_exception():
    raise SMTPException
...
class TestErrorHandling(WagtailTestUtils, TransactionTestCase):
    ...
    def test_smtp_exception(self):
        ...
        with self.assertRaises(SMTPException):
            something_that_i_am_testing_that_raises_smtp_exception()
```

Here is an example where you can see it in practice:
https://github.com/neon-jungle/wagtail-birdsong/pull/36/files#diff-041ae52229a7d4829a12f2c872c197a22146eaee3c02ea522d940d6fb886998fR23-R24

However, as said in the begining. Don't just `raise` an exception. Simulate it with test data and function calls.

**Comment by david-smejkal** [comment]
What you want to do is to simulate the situation where the code throws such an exception in your test by supplying specific test data and calling core birdsong functions. That's because you want to figure out whether the exception is an acceptable situation or whether it's something that should be remedied in some way.

Now to answer your question you can "throw" and exception forcefully with the `raise` keyword and then test for it as follows:
```python
from smtplib import SMTPException
...
def something_that_i_am_testing_that_raises_smtp_exception():
    raise SMTPException
...
class TestErrorHandling(WagtailTestUtils, TransactionTestCase):
    ...
    def test_smtp_exception(self):
        ...
        with self.assertRaises(SMTPException):
            something_that_i_am_testing_that_raises_smtp_exception()
```

@danhayden, here is an example where you can see testing for an expected exception in practice:
https://github.com/neon-jungle/wagtail-birdsong/pull/36/files#diff-041ae52229a7d4829a12f2c872c197a22146eaee3c02ea522d940d6fb886998fR23-R24

However, as said in the begining. Don't just `raise` an exception. Simulate it with test data and core birdsong function calls.

**Comment by danhayden** [comment]
Still no tests for this but I don't think that should prevent this small fix being merged. Let me know if any changes are required.

**Pull request merged**